File: day06/python/part1.py
#!/usr/bin/env python3

import helper

SIZE = 4


def process(text: str) -> int:
    for i in range(len(text) - SIZE + 1):
        sub = text[i : i + SIZE]
        if len(sub) == len(set(sub)):
            return i + SIZE
        #
    #
    return -1


def main():
    ex1 = "mjqjpqmgbljsphdztnvjfqwrcgsmlb"  # 7
    ex2 = "bvwbjplbgvbhsrlpgdmjqwftvncz"  # 5
    ex3 = "nppdvjthqldpwncqszvftbrmjlhg"  # 6
    ex4 = "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg"  # 10
    ex5 = "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw"  # 11

    text = helper.read("input.txt")
    result = process(text)
    print(result)


##############################################################################

if __name__ == "__main__":
    main()

File: day06/golang/part1/part1.go
package main

import (
	"aoc/lib/jfile"
	"fmt"
)

const SIZE = 4

func uniq(text string) bool {
	d := make(map[rune]struct{})
	for _, c := range text {
		_, found := d[c]
		if found {
			return false
		}
		// else
		d[c] = struct{}{}
	}
	return true
}

func process(text string) int {
	for i := 0; i < len(text)-SIZE+1; i++ {
		sub := text[i : i+SIZE]
		if uniq(sub) {
			return i + SIZE
		}
	}
	return -1
}

func main() {
	// ex := "mjqjpqmgbljsphdztnvjfqwrcgsmlb" // 7
	// ex := "bvwbjplbgvbhsrlpgdmjqwftvncz" // 5
	// ex := "nppdvjthqldpwncqszvftbrmjlhg" // 6
	// ex := "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg" // 10
	// ex := "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw" // 11

	// text := ex

	text, _ := jfile.Read("input.txt")
	result := process(text)
	fmt.Println(result)
}
